Assistant: add project tree tool

sharon-wang opened the pull request:
Initial service scaffolding and project tree tool for https://github.com/posit-dev/positron/issues/7110.

Looking for feedback on
- structure of the changes -- thoughts on having a workspace-specific service for assistant tools?
- how is the experience with questions around project structure? Would a more specific tool be helpful? Any other workspace-related tools that might be useful?

Upcoming
- tools to get info about specific files in the workspace
- other workspace-related tools will be added to the service scaffolded in this PR

### QA Notes

Questions about project structure, number of files/folders, insight about the nature of the project, etc. should now leverage the project structure tool call.

Questions about the content within specific files or searching for specific content across files/directories seem to work to a reasonable extent, but were not the focus of this work.

sharon-wang committed:
add project tree tool to assistant

sharon-wang committed:
use array-based representation of project tree

a bit more efficient on tokens, while still maintaining well-formed JSON structure

sharon-wang committed:
move project tree construction into positron core

sharon-wang committed:
fix up tree construction and promise handling

sharon-wang committed:
reduce token output by ignoring files and not including 'isexcluded' metadata

file/dir - specific info can be a separate tool call

seeM reviewed (commented):
For tools like `getProjectTree` that just wrap an extension API method that only exists for this specific case, what do you think about moving this tool into Positron core? Like the edit file tool here: https://github.com/posit-dev/positron/blob/main/src/vs/workbench/contrib/chat/common/tools/editFileTool.ts

EDIT: It might also mean that some of the workspace-specific service could be inlined into the tool implementation.

sharon-wang committed:
move project tool to built-in tools

sharon-wang committed:
🧹 cleanup

sharon-wang commented on src/vs/workbench/contrib/chat/common/tools/projectTreeTool.ts:
```diff
@@ -0,0 +1,139 @@
+import { CancellationToken } from '../../../../../base/common/cancellation.js';
+import { localize } from '../../../../../nls.js';
+import { IWorkspaceContextService } from '../../../../../platform/workspace/common/workspace.js';
+import { IExplorerService } from '../../../files/browser/files.js';
```
I was able to move the tool over, but I have a layering issue on this import now.

```
Bad layering. You are not allowed to access browser from here, allowed layers are: [common]eslint[local/code-layering](https://github.com/microsoft/vscode/wiki/Source-Code-Organization)

Imports violates 'vs/base/common/** or vs/base/parts/*/common/** or vs/platform/*/common/** or vs/editor/common/** or vs/editor/contrib/*/common/** or vs/workbench/common/** or vs/workbench/services/*/common/** or vs/workbench/contrib/*/common/** or vs/workbench/contrib/terminal/terminalContribChatExports* or vs/workbench/contrib/terminal/terminalContribExports* or vscode-notebook-renderer or @vscode/tree-sitter-wasm or vs/nls.js or vs/amdX.js' restrictions. See https://github.com/microsoft/vscode/wiki/Source-Code-Organizationeslint[local/code-import-patterns](https://github.com/microsoft/vscode/wiki/Source-Code-Organization)
```

seeM replied:
Would it work to move this file to: `src/vs/workbench/contrib/chat/browser/tools/projectTreeTool.ts`?

If needed we could also create and register our own `PositronBuiltinToolsContribution`.

sharon-wang replied:
yes it does! I ended up registering our own `PositronBuiltinToolsContribution` as well 👍

seeM commented on src/vs/workbench/contrib/chat/common/tools/projectTreeTool.ts:
```diff
@@ -0,0 +1,139 @@
+
+		return {
+			content: workspaceTrees.map(dirTree => ({ kind: 'text', value: JSON.stringify(dirTree) })),
+			toolResultMessage: treeErrors.length > 0 ? treeErrors.join('\n') : undefined
```
How are you thinking about when to include text in `content` versus `toolResultMessage` versus throwing an error?

sharon-wang replied:
I've now modified the error handling to throw, as it's pretty unexpected if we can't generate the project tree for some reason. Instead of returning errors in the toolResultMessage here, we throw just prior if any errors occur.

seeM commented on src/vs/workbench/contrib/chat/common/tools/projectTreeTool.ts:
```diff
@@ -0,0 +1,139 @@
+	}
+}
+
+const IGNORED_DIRECTORIES = [
```
This can be a future improvement: I think there is existing functionality in the editor for ignoring files in certain contexts, which may also use `.gitignore`. Would we want to hook into that?

sharon-wang replied:
Yes, I think it would be helpful to hook into that! Maybe an optional input parameter to indicate whether the project tree should be constructed with or without gitignored files? I added a TODO note inline. There's also the `files.exclude` setting that we likely want to read as well.

I could see a user being confused about missing files and the "full" project tree could help reveal which files are hidden. At the same time, probably for most use cases, we don't need to include the ignored files.

seeM reviewed (commented):
Nice!

* I added a few more general thoughts about tool descriptions.
* I'm happy with this approach to adding tools which require functionality where it doesn't make sense to expand the extension API. I'm not sure if we'll run into any limitations with these living in core code instead of the extension, but we can see how it goes.
* I think this is a good starting point for workspace tools. Will try it out once the layering issue is resolved. Looks good to merge once we're happy with the initial prompts.

sharon-wang committed:
create positron builtin tools contribution

this helps resolve project tools layering issues and also sets us up for future positron tools

sharon-wang committed:
add detail to project tree model description

sharon-wang committed:
update model description and error handling

sharon-wang committed:
🧹 cleanup

sharon-wang commented on src/vs/workbench/contrib/chat/common/tools/projectTreeTool.ts:
```diff
@@ -0,0 +1,139 @@
+export const ProjectTreeToolData: IToolData = {
+	id: InternalProjectTreeId,
+	displayName: localize('chat.tools.getProjectTree', "Get Project Tree"),
+	modelDescription: `Construct a tree of the directories and files in the project.`,
```
This is really helpful feedback, thank you! I've rewritten the modelDescription with more detail. It does seem like it will try to invoke the project tree tool only once now, whereas it would invoke the tool every time without the extra detail in the description. I can likely further refine the description, but I think it's at a decent spot for now!

seeM reviewed (approved):
LGTM! I haven't had a chance to try it out yet, but think we can merge. Will try it out and share feedback separately tomorrow
